### ISSUE TITLE
A problem of tff.learning. build _ federated _ averaging _ process?

### ISSUE THREAD

**Issue by Chun78HunZi** [tensorflow/federated#951]
Why is the default parameter of server_optimizer_fn in TFF. learning. build _ federated _ averaging _ process SGD instead of using weighted average as mentioned in the paper

**Comment by ZacharyGarrett** [comment]
Correct that `tff.learning.build_federated_averaging_process` isn't implemented exactly as averaging of the client model weights. However it is effectively the same algorithm when using the default server optimizer parameter of `tf.keras.optimizers.SGD(learning_rate=1.0)`.

Instead, the clients send back _the difference_ between their final model and the initial model (client "model_update"), rather than the weights themselves. This helps with things like lossy compression techniques, as smaller magnitude values encode better and/or with less error. 

These updates are averaged, then applied to the server as a pseudo-gradient using the server optimizer. When the server optimizer is SGD with a learning rate of `1.0` this recovers the same semantics as averaging the model parameters themselves. Loosely, if we take `W_i` as the global model at the start of round `i`, and `W_j` is a client `j`'s model after local training, the model round `i+1` is shown as:

`W_i+1 = W_i + 1.0 * Average_j(W_j - W_i) = Average_j(W_j)`

By implementing this way, its easy to experiment with more sophisticated algorithms for optimization on the server. See https://arxiv.org/abs/2003.00295 for some examples.
